### PR TITLE
fix syncWith:transform:revert  implementation

### LINKED ISSUE #33: syncWith:transform:revert API Error
When I syncWith   EZRNode<NSNumber *>  and  EZRNode<NSString *>  ,there will be a mistake
   
```
    EZRMutableNode<NSNumber *> *nodeA = [EZRMutableNode new];
    EZRMutableNode<NSString *> *nodeB = [EZRMutableNode new];
    
    [nodeA syncWith:nodeB transform:^id _Nonnull(NSNumber * _Nonnull source) {
        NSLog(@"class is %@",[source class]);    // source is not NSNumber but  __NSCFConstantString
         return @"9";
    } revert:^NSNumber * _Nonnull(id  _Nonnull target) {
         return @2;
    }];
```

The current API is

`- (id<EZRCancelable>)syncWith:(EZRNode *)othEZRNode transform:(id (^)(T source))transform revert:(T (^)(id target))revert;`

Maybe this way is right

`- (id<EZRCancelable>)syncWith:(EZRNode<T> *)othEZRNode transform:(T (^)(T source))transform revert:(T (^)(T target))revert;`

### PR DESCRIPTION
**Pull request by zhudaye12138**
for  issues   [](https://github.com/meituan/EasyReact/issues/33)

fixes #33  syncWith:transform:revert implementation  and associated UT

### CODE REVIEW

File: Example/Tests/EZRNode+Operation.m [hunk 1]
```diff
@@ -514,19 +517,19 @@
...
             EZRMutableNode<NSNumber *> *value1 = EZRMutableNode.new;
-            EZRMutableNode<NSNumber *> *value2 = EZRMutableNode.new;
+            EZRMutableNode<NSString *> *value2 = EZRMutableNode.new;
             
-            [value1 syncWith:value2 transform:^NSNumber *(NSNumber *source) {
-                return @(source.unsignedIntegerValue  * 2);
-            } revert:^NSNumber *(NSNumber *target) {
-                return @(target.unsignedIntegerValue  / 2);
+            [value1 syncWith:value2 transform:^NSString* _Nonnull(NSNumber * _Nonnull source) {
+                return @(source.integerValue*2).stringValue;
```
Comment (WilliamZang) [inline Example/Tests/EZRNode+Operation.m]: `integerValue*2 -> integerValue * 2`

add space.

File: EasyReact/Classes/Core/EZRNode+Operation.m [hunk 2]
```diff
@@ -170,8 +170,8 @@ - (EZRNode *)delayOnMainQueue:(NSTimeInterval)timeInterval {
     EZRMapTransform *mapTransform = [[EZRMapTransform alloc] initWithMapBlock:transform];
     EZRMapTransform *mapRevert = [[EZRMapTransform alloc] initWithMapBlock:revert];
     
-    id<EZRCancelable> transformCancelable = [self linkTo:othEZRNode transform:mapTransform];
-    id<EZRCancelable> revertCancelable = [othEZRNode linkTo:self transform:mapRevert];
+    id<EZRCancelable> transformCancelable = [othEZRNode linkTo:self transform:mapTransform];
```
Comment (WilliamZang) [inline EasyReact/Classes/Core/EZRNode+Operation.m]: fix typo by the way. `othEZRNode -> otherNode`

**Comment by WilliamZang** [comment]
Please squash commit to a single commit. See [git-scm book](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_squashing).

File: EasyReact/Classes/Core/EZRNode+Operation.m (hunk 2 above)
Comment (ValiantCat) [reply EasyReact/Classes/Core/EZRNode+Operation.m]: also check header file

File: Example/Tests/EZRNode+Operation.m [hunk 3]
```diff
@@ -487,7 +487,9 @@
             expect(value2.hasUpstreamNode).to(beTrue());
             expect(value1.hasDownstreamNode).to(beTrue());
             expect(value2.hasDownstreamNode).to(beTrue());
-            expect(value1.value).to(equal(@2));
+
```
Comment (ValiantCat) [inline Example/Tests/EZRNode+Operation.m]: Can not expression value1's value is 2, or value2's value is 1.

File: Example/Tests/NSObject+ER_Extension.m [hunk 4]
```diff
@@ -128,8 +128,8 @@
         obj2.property3 = @"yyy";
         
         EZR_PATH(obj2, property3) = EZR_PATH(obj1, property1);
-        expect(obj1.property1).to(equal(@"xxx"));
-        expect(obj2.property3).to(equal(@"xxx"));
+        expect(obj1.property1).to(equal(@"yyy"));
```
Comment (WilliamZang) [inline Example/Tests/NSObject+ER_Extension.m]: This testcase is correctly, but implementation is wrong with sync direct.
See [EZRPathTrampoline.m line71](https://github.com/meituan/EasyReact/blob/2.2.1/EasyReact/Classes/Utils/EZRPathTrampoline.m#L71)

Comment (WilliamZang) [inline Example/Tests/NSObject+ER_Extension.m]: This test case is correctly, but implementation is wrong with sync direct.
See [EZRPathTrampoline.m line71](https://github.com/meituan/EasyReact/blob/2.2.1/EasyReact/Classes/Utils/EZRPathTrampoline.m#L71)

File: Example/Tests/EZRNode+Operation.m (hunk 3 above)
Comment (WilliamZang) [reply Example/Tests/EZRNode+Operation.m]: You should give a default direction when syncWith: be used.

File: Example/Tests/EZRNode+Operation.m [hunk 5]
```diff
@@ -501,7 +503,8 @@
             EZRMutableNode *value2 = [EZRMutableNode value:@2];
             
             id<EZRCancelable> cancelable = [value1 syncWith:value2];
-            expect(value1.value).to(equal(@2));
+            value2.value = @4;
```
Comment (WilliamZang) [inline Example/Tests/EZRNode+Operation.m]: You should give a default direction when syncWith: be used.

File: Example/Tests/EZRNode+Operation.m (hunk 3 above)
Comment (zhudaye12138) [reply Example/Tests/EZRNode+Operation.m]: ok

File: Example/Tests/NSObject+ER_Extension.m (hunk 4 above)
Comment (zhudaye12138) [reply Example/Tests/NSObject+ER_Extension.m]: ok, I got it

**Comment by WilliamZang** [comment]
@zhudaye12138 There are 2 commits in your PR. You have to squash to a single commit. Follow those steps if you don't know how to do:

1. Make sure your fork repository's master branch is up-to-date. 
```shell
cd path-to-your-local-repository
git checkout syncError
git branch -D master
git pull https://github.com/meituan/EasyReact.git master:master
```
2. Rebuild your development branch which is 'optimizeCode' base on your master branch.
```shell
git checkout master
git branch -D syncError
git checkout -b syncError
git cherry-pick 01a0e6044d503c94625becbd5dc074fdaa6356e8
```
3. Change your last commit message, copy-paste whose three line together:
```shell
git commit --amend -m "fix : syncWith:transform:revert link directs error

[#33]"
``` 
4. Push your branch again use `push -f`.

**Comment by zhudaye12138** [comment]
NO problem .  I've been too busy these two days, forget to log in github.   

@WilliamZang  Thank you very much！

**Pull request merged**